crypt-gost-yescrypt.c

```diff
@@ -0,0 +1,182 @@
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "crypt-port.h"
+
+#if INCLUDE_gost_yescrypt
+
+#include "crypt-private.h"
+#include "byteorder.h"
+
+#include <errno.h>
+
+#define YESCRYPT_INTERNAL
+#include "alg-yescrypt.h"
+#undef YESCRYPT_INTERNAL
+
+#include "alg-gost3411-2012-hmac.h"
+
+/* upper level hmac for tests */
+#ifndef outer_gost_hmac256
+#define outer_gost_hmac256 gost_hmac256
+#endif
+
+
+/* For use in scratch space by crypt_gost_yescrypt_rn().  */
+typedef struct
+{
+  yescrypt_local_t local;
+  gost_hmac_256_t gostbuf;
+  uint8_t outbuf[CRYPT_OUTPUT_SIZE];
+  uint8_t gsetting[CRYPT_OUTPUT_SIZE];
+  uint8_t hk[32], interm[32], y[32];
+  uint8_t *retval;
+} crypt_gost_yescrypt_internal_t;
+
+static_assert (sizeof (crypt_gost_yescrypt_internal_t) <= ALG_SPECIFIC_SIZE,
+               "ALG_SPECIFIC_SIZE is too small for GOST-YESCRYPT.");
+
+/*
+ * As OUTPUT is initialized with a failure token before gensalt_yescrypt_rn
+ * is called, in case of an error we could just set an appropriate errno
+ * and return.
+ */
+void
+gensalt_gost_yescrypt_rn (unsigned long count,
+                          const uint8_t *rbytes, size_t nrbytes,
+                          uint8_t *output, size_t o_size)
+{
+  if (o_size < 4 + 8 * 6 + BASE64_LEN (nrbytes) + 1 ||
+      CRYPT_GENSALT_OUTPUT_SIZE < 4 + 8 * 6 + BASE64_LEN (nrbytes) + 1)
+    {
+      errno = ERANGE;
+      return;
+    }
+
+  /* We pass 'o_size - 1' to gensalt, because we need to shift
+           the prefix by 1 char to insert the gost marker.  */
+  gensalt_yescrypt_rn (count, rbytes, nrbytes, output, o_size - 1);
+
+  /* Check for failures.  */
+  if (output[0] == '*')
+    return;
+
+  /* Shift output one byte further.  */
+  memmove (output + 1, output, strlen ((const char *) output) + 1);
+
+  /* Insert the gost marker.  */
+  output[1] = 'g';
+}
+
+void
+crypt_gost_yescrypt_rn (const char *phrase, size_t phr_size,
+                        const char *setting, size_t set_size,
+                        uint8_t *output, size_t o_size,
+                        void *scratch, size_t s_size)
+{
+  if (o_size < set_size + 1 + 43 + 1 ||
+      CRYPT_OUTPUT_SIZE < set_size + 1 + 43 + 1 ||
+      s_size < sizeof (crypt_gost_yescrypt_internal_t))
+    {
+      errno = ERANGE;
+      return;
+    }
+
+  /* Fail when called with wrong prefix.  */
+  if (strncmp (setting, "$gy$", 4))
+    {
+      errno = EINVAL;
+      return;
+    }
+
+  crypt_gost_yescrypt_internal_t *intbuf = scratch;
+
+  if (yescrypt_init_local (&intbuf->local))
+    return;
+
+  /* convert gost setting to yescrypt setting */
+  intbuf->gsetting[0] = '$';
+  intbuf->gsetting[1] = 'y';
+  intbuf->gsetting[2] = '$';
+  XCRYPT_STRCPY_OR_ABORT (&intbuf->gsetting[3], set_size - 3, setting + 4);
```

Comment:
Just side note: `XCRYPT_STRCPY_OR_ABORT` aborts only if `assert` is defined, which relies on `crypt-port.h` included first. Which relies on the convention for include order and not on the code.

Reply:
`XCRYPT_STRCPY_OR_ABORT` is actually defined in `crypt-port.h`, to make sure all requirements are met.